File: Napakalaki Ruby/lib/napakalaki.rb
# encoding: UTF-8

# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative 'player'
require_relative 'card_dealer'
require_relative 'cultist_player'

module Napakalaki

  require "singleton"

  class Napakalaki

    include Singleton

    def initPlayers (names)

      @dealer = CardDealer.instance
      @players = Array.new

      for i in 0..(names.size-1) do

        @players.push(Player.new(names[i]))

      end

      @indexCurrentPlayer = -1
      
    end


    def nextPlayer

      if (@indexCurrentPlayer == -1) then

          rand = Random.new

          @indexCurrentPlayer = rand(0..(@players.size-1))    

      else

        @indexCurrentPlayer = ((@indexCurrentPlayer + 1) % @players.size)

      end
      
      @currentPlayer = @players[@indexCurrentPlayer]

      return @currentPlayer

    end


    def nextTurnAllowed

      if @currentPlayer == nil then

              return true

      else

        return @currentPlayer.validState

      end

    end


    def developCombat

      result = @currentPlayer.combat(@currentMonster)

      @dealer.giveMonsterBack(@currentMonster)
      
      if result == CombatResult::LOSEANDCONVERT then
        
        cultist = @dealer.nextCultist
        cultistPLayer = CultistPlayer.new(@currentPlayer, cultist)
        
        @players.delete(@currentPlayer)
        @players.insert(@indexCurrentPlayer, cultistPLayer)
        @currentPlayer = cultistPLayer
        
      end

      return result

    end


    def discardVisibleTreasures (treasures)

      for i in 0..(treasures.size-1) do

        @currentPlayer.discardVisibleTreasure(treasures[i])
        @dealer.giveTreasureBack(treasures[i])

      end

    end


    def discardHiddenTreasures (treasures)

      for i in 0..(treasures.size-1) do

        @currentPlayer.discardHiddenTreasure(treasures[i])
        @dealer.giveTreasureBack(treasures[i])

      end

    end


    def makeTreasuresVisible (treasures)

      for i in 0..(treasures.size-1) do

        @currentPlayer.makeTreasureVisible(treasures[i])

      end

    end


    def buyLevels (visibleTreasures, hiddenTreasures)

      return @currentPlayer.buyLevels(visibleTreasures,hiddenTreasures);

    end


    def initGame (players)

      initPlayers(players)
      @dealer.initCards
      nextTurn

    end


    def getCurrentPlayer

      return @currentPlayer

    end


    def getCurrentMonster

      return @currentMonster

    end


    def nextTurn
      
      if nextTurnAllowed then

        nextPlayer

        @currentMonster = @dealer.nextMonster

        if @currentPlayer.isDead then

          @currentPlayer.initTreasures
          
        end

        return true

      else

        return false

      end

    end


    def endOfGame (combatResult)

      return (combatResult == CombatResult::WINANDWINGAME)

    end

  end

end

File: Napakalaki Java/src/NapakalakiGame/Napakalaki.java
/*
 */
package NapakalakiGame;

import java.util.ArrayList;
import java.util.Random;

/**
 *
 *
 */
public class Napakalaki {
    
    private static final Napakalaki instance = new Napakalaki();
    
    private ArrayList<Player> players = new ArrayList();
    private Player currentPlayer;
    private Monster currentMonster;
    private int indexCurrentPlayer = -1;
    
    private Napakalaki() {}
    
    public static Napakalaki getInstance() {
        
        return instance;
    }
    
    private void initPlayers(ArrayList<String> nam){
        
        for(int i=0; i<nam.size() ; i++)
            
            players.add(new Player(nam.get(i)));
        
    }
    
    private Player nextPlayer(){
        
        if (indexCurrentPlayer == -1){
            
            Random rand = new Random();
            
            indexCurrentPlayer = rand.nextInt(players.size());            
        }
        else
        
            indexCurrentPlayer = (indexCurrentPlayer + 1) % players.size();
            
        currentPlayer = players.get(indexCurrentPlayer);   
        
        return currentPlayer;
        
    }
    
    private boolean nextTurnAllowed(){
        
        if (currentPlayer == null)
            
            return true;
        
        else
            
            return currentPlayer.validState();
    }
    
    public CombatResult developCombat(){
    
        CombatResult result;
        
        CardDealer dealer = CardDealer.getInstance();
        
        result = currentPlayer.combat(currentMonster);
        
        if(result == CombatResult.LoseAndConvert){
            
            Cultist cultist = dealer.nextCultist();
            
            CultistPlayer cp = new CultistPlayer(currentPlayer, dealer.nextCultist());
            
            currentPlayer = cp;
            
            players.set(indexCurrentPlayer, currentPlayer);
               
        }
        
        dealer.giveMonsterBack(currentMonster);
        
        return result;
        
    }
    
    public void discardVisibleTreasures(ArrayList<Treasure> treasures){
        
        CardDealer dealer = CardDealer.getInstance();
        
        for (int i = 0; i<treasures.size(); i++){
        
            currentPlayer.discardVisibleTreasure(treasures.get(i));
            dealer.giveTreasureBack(treasures.get(i));
            
        }
            
    }
    
    public void discardHiddenTreasures(ArrayList<Treasure> treasures){   
        
        CardDealer dealer = CardDealer.getInstance();
        
        for (int i = 0; i<treasures.size(); i++){
        
            currentPlayer.discardHiddenTreasure(treasures.get(i));
            dealer.giveTreasureBack(treasures.get(i));
            
        }
    }
    
    public void makeTreasuresVisible(ArrayList<Treasure> treasures){
        
        for (int i = 0; i < treasures.size(); i++)
            
            currentPlayer.makeTreasureVisible(treasures.get(i));
        
    }
    
    public boolean buyLevels(ArrayList<Treasure> visible, ArrayList<Treasure> hidden){
        
        return currentPlayer.buyLevels(visible,hidden);
    }
    
    public void initGame(ArrayList<String> nam){
        
        CardDealer dealer = CardDealer.getInstance();
        
        dealer.initCards();
        initPlayers(nam);
        nextTurn();
    }
    
    public Player getCurrentPlayer(){
       
        return currentPlayer;
    }
    
    public Monster getCurrentMonster(){
        
        return currentMonster;
    }
    
    public boolean nextTurn(){
        
        CardDealer dealer = CardDealer.getInstance();
        
        if (nextTurnAllowed()){
            
            nextPlayer();
            
            currentMonster = dealer.nextMonster();
            
            if (currentPlayer.isDead())
                
                currentPlayer.initTreasures();
            
            return true;
        }
        
        else
            
            return false;
        
    }
    
    public boolean endOfGame(CombatResult cr){
        
        return (cr == CombatResult.WinAndWinGame);
        
    }
 
}
